lib/factbase/lazy_taped.rb

```diff
@@ -0,0 +1,294 @@
+      root_pairs[n.object_id] = m.object_id
+      # rubocop:enable Lint/HashCompareByIdentity
+      root_origin << n
+    end
+    root.instance_variable_set(:@copied, true)
+    # Also populate this instance's @origin if it's different from root
+    sync_origin_from_root unless root.equal?(self)
+  end
+
+  # Returns the copied map for a given original object_id.
+  def get_copied_map(original_oid)
+    root_instance.instance_variable_get(:@original_to_copy)[original_oid]
+  end
+
+  # Decorator of Hash that defers copy until write.
+  class LazyTapedHash
+    decoor(:origin)
+
+    def initialize(origin, added, lazy_taped)
+      @origin = origin
+      @origin_oid = origin.object_id
+      @added = added
+      @lazy_taped = lazy_taped
+    end
+
+    def keys
+      current_map.keys
+    end
+
+    def map(&)
+      current_map.map(&)
+    end
+
+    def [](key)
+      v = current_map[key]
+      v = LazyTapedArray.new(v, key, @origin_oid, @added, @lazy_taped) if v.is_a?(Array)
+      v
+    end
+
+    def []=(key, value)
+      @lazy_taped.ensure_copied!
+      map = @lazy_taped.get_copied_map(@origin_oid)
+      map[key] = value
+      @added.append(map.object_id)
+    end
+
+    private
+
+    def current_map
+      @lazy_taped.copied? ? @lazy_taped.get_copied_map(@origin_oid) : @origin
+    end
+  end
+
+  # Decorator of Array that defers copy until write.
+  class LazyTapedArray
+    def initialize(origin, key, map_oid, added, lazy_taped)
+      @origin = origin
+      @key = key
+      @map_oid = map_oid
+      @added = added
+      @lazy_taped = lazy_taped
+    end
+
+    def each(&)
+      return to_enum(__method__) unless block_given?
+      current_array.each(&)
+    end
+
+    def [](idx)
+      current_array[idx]
+    end
+
+    def to_a
+      current_array.to_a
+    end
+
+    def any?(&)
+      current_array.any?(&)
+    end
+
+    def <<(item)
+      @lazy_taped.ensure_copied!
+      map = @lazy_taped.get_copied_map(@map_oid)
+      map[@key] << item
+      @added.append(map.object_id)
+    end
+
+    def uniq!
+      @lazy_taped.ensure_copied!
+      map = @lazy_taped.get_copied_map(@map_oid)
+      map[@key].uniq!
+      @added.append(map.object_id)
+    end
+
+    private
+
+    def current_array
+      if @lazy_taped.copied?
+        map = @lazy_taped.get_copied_map(@map_oid)
+        map[@key]
+      else
+        @origin
+      end
+    end
+  end
+
+  private
+
+  def root_instance
+    @root || self
+  end
+
+  def root?
+    @root.nil?
+  end
+
+  def root_origin
+    root_instance.instance_variable_get(:@origin)
+  end
+
+  def root_inserted
+    root_instance.instance_variable_get(:@inserted) || @inserted || []
+  end
+
+  def root_deleted
+    root_instance.instance_variable_get(:@deleted) || @deleted || []
+  end
+
+  def root_added
+    root_instance.instance_variable_get(:@added) || @added || []
+  end
+
+  def sync_origin_from_root
+    # Find maps in root's @origin that correspond to our @original
```

Comment:
@HaidarJbeily7 do we need this comment? If it is so, it's better to rewrite it to make it answer the question "why?"

https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/